cql3/statements/create_view_statement.cc

```diff
@@ -260,6 +260,25 @@ view_ptr create_view_statement::prepare_view(data_dictionary::database db) const
         throw exceptions::invalid_request_exception(format("Must select at least a column for a Materialized View"));
     }
 
+    // Cassandra requires that if CLUSTERING ORDER BY is used, it must specify
+    // all clustering columns. Scylla relaxes this requirement and allows just
+    // a subset of the clustering columns. But it doesn't make sense (and it's
+    // forbidden) to list something which is not a clustering key column in
+    // the CLUSTERING ORDER BY. Let's verify that:
```

Comment:
With `CREATE TABLE` Scylla is also pretty strict about what is allowed.
We only allow full clustering prefixes there.

```cql
-- The same colum twice - disallowed
cqlsh:ks> CREATE TABLE test_table (pk int, c1 int, c2 int, c3 int, PRIMARY KEY (pk, c1, c2, c3)) WITH CLUSTERING ORDER BY (c1 ASC, c1 ASC);
InvalidRequest: Error from server: code=2200 [Invalid query] message="Missing CLUSTERING ORDER for column c2"
-- Missing c2 - not a valid clustering prefix - disallowed
cqlsh:ks> CREATE TABLE test_table (pk int, c1 int, c2 int, c3 int, PRIMARY KEY (pk, c1, c2, c3)) WITH CLUSTERING ORDER BY (c1 ASC, c3 ASC);
InvalidRequest: Error from server: code=2200 [Invalid query] message="Missing CLUSTERING ORDER for column c2"
-- Wrong order - c2, c1 - not a valid clustering prefix - disallowed
cqlsh:ks> CREATE TABLE test_table (pk int, c1 int, c2 int, c3 int, PRIMARY KEY (pk, c1, c2, c3)) WITH CLUSTERING ORDER BY (c2 ASC, c1 ASC);
InvalidRequest: Error from server: code=2200 [Invalid query] message="The order of columns in the CLUSTERING ORDER directive must be the one of the clustering key (c1 must appear before c2)"
-- A valid clustering prefix - allowed
cqlsh:ks> CREATE TABLE test_table (pk int, c1 int, c2 int, c3 int, PRIMARY KEY (pk, c1, c2, c3)) WITH CLUSTERING ORDER BY (c1 ASC, c2 ASC);

```

I think the validation should be the same for normal tables and materialized views.  Why is something allowed on one of them, but not the other one. Materialized views are tables after all.

Reply:
It would be easy to allow specifying them in any order, but I'm not convinced that it's a good idea.
https://github.com/scylladb/scylladb/blob/ef50170120153b7cc11bea7cfccfa85b5a28bd82/cql3/statements/select_statement.cc#L1847-L1854

Reply:
> With `CREATE TABLE` Scylla is also pretty strict about what is allowed. We only allow full clustering prefixes there.
> I think the validation should be the same for normal tables and materialized views. Why is something allowed on one of them, but not the other one. Materialized views are tables after all.

You're right that the validation in CREATE TABLE is much stricter.

Until a minute ago I was sure that this was just laziness on the Cassandra programmer's side, but now I realize it actually serves an important purpose (which should be documented): If PRIMARY KEYS defines clustering keys c1, c2 in this order, a CLUSTERING ORDER BY cannot change this order! Allowing "CLUSTERING ORDER BY (c2 asc, c1 asc)" might suggest to someone that this reverses the priority between the two clustering keys, when it can't!

That being said, this (the leniency of the order of the column list in ORDER BY in MV) is a pre-existing issue, which this patch shouldn't solve. This patch only gives an error when the given columns do not make sense at all (they aren't clustering columns - possibly even don't exist at all). I'll open a new issue to reconsider the leniency. That issue will have backward compatibility considerations (CQL code which current Scylla accepts and makes sense and we'll want to forbid). But this patch is simpler and less controversial.

Reply:
I think I agree with you, and opened https://github.com/scylladb/scylladb/issues/12936.
Maybe it makes sense, though, to commit this patch first and then think about making it even more strict?